### ISSUE TITLE
Hangs waiting for backgrounded SSH master process to exit

### ISSUE THREAD

**Issue by joshtriplett** [sociomantic-tsunami/git-hub#188]
With SSH configured as follows:

```
Host *
    ControlPath ~/.ssh/master-%r@%h:%p
    ControlMaster auto
    ControlPersist 60
```

...commands like `git hub pull new` will hang on the `Pushing <branch> to <branch> in fork` stage until the backgrounded master process exits, rather than just waiting for the SSH connection itself to terminate.  `ps auxf` shows the git process defunct, but the ssh master still running; as soon as the master exits (after 60 seconds with this configuration), the `git hub` command proceeds.

**Comment by llucax** [comment]
Mmm, inside this tool we just call git commands, I don't think there is much we can do about this. If you want you can use `git hub -v` to see exactly which command is hanging and try to run it manually. Please let me know if that still hangs, so I can close this issue as invalid.

**Comment by joshtriplett** [comment]
Git commands don't hang; they terminate and return to the shell.  When they spawn ssh (as `git push` does), ssh forks a master-mode process into the background.  `git hub` seems to be waiting for that ssh master-mode process to terminate, rather than only the git process.

**Comment by jwilk** [comment]
Very odd.
I can reproduce it, but not every time.
This is what `gdb` attached to the hanged process says:
```
(gdb) py-bt
Traceback (most recent call first):
  <built-in method poll of select.poll object at remote 0xf6f184e0>
  File "/usr/lib/python2.7/subprocess.py", line 1471, in _communicate_with_poll
    ready = poller.poll()
  File "/usr/lib/python2.7/subprocess.py", line 1417, in _communicate
    stdout, stderr = self._communicate_with_poll(input)
  File "/usr/lib/python2.7/subprocess.py", line 800, in communicate
    return self._communicate(input)
  File "/usr/bin/git-hub", line 215, in git
    (stdout, stderr) = proc.communicate()
  File "/usr/bin/git-hub", line 239, in git_quiet
    return git(*args, **kwargs)
  File "/usr/bin/git-hub", line 249, in git_push
    git_quiet(1, *cmd, **kwargs)
  File "/usr/bin/git-hub", line 1273, in push
    force=force)
  File "/usr/bin/git-hub", line 1388, in run
    force=args.force_push)
  File "/usr/bin/git-hub", line 621, in check_config_and_run
    cmd.run(parser, args)
  File "/usr/bin/git-hub", line 2037, in main
    args.run(parser, args)
  File "/usr/bin/git-hub", line 2043, in <module>
    main()
```

**Comment by llucax** [comment]
Mmm, that's weird, we are just using Python's `subprocess` module, which AFAIK it just wait on the process that was forked, we don't have any information about your SSH process in this tool. In fact, I'm using also the `ControlMaster auto` option in SSH (although I don't use the `ControlPersist` option) and never had problems.

Can you maybe run a `strace` to see which process is waiting on which. Because for sure in the git-hub tool we are not waiting for anything but the `git` process we are invoking (as I said, we don't even know or care about the existence of SSH or any other method `git` uses to communicate with the world), so if it's not the `git` process that's waiting for the master SSH process, then maybe it is some issue with Python itself (BTW, can you mention your Python's version?), but again I don't know how Python can know anything about the underlaying SSH process git is forking.

**Comment by llucax** [comment]
Maybe there is some environment change between when running `git` inside Python and directly in the shell that's changing git's internal behavior.

**Comment by jwilk** [comment]
The problem appears to be that the daemonized SSH process didn't close stderr.
This stderr is a pipe to git-hub, so git-hub waits for the input until the ssh process quits...

**Comment by jwilk** [comment]
Minimal reproducer in shell:
```
ssh ssh.example.org echo HANG... 2>&1 | cat
```
This looks very much like an SSH bug, but I wonder why is git-hub capturing stderr in the first place.

**Comment by llucax** [comment]
Good catch! I don't remember the exact original reason why I ended up doing this, but I'm capturing stderr as another way to detect errors in git (I don't remember if I actually came across some situation when git encountered some error but the status code returned was 0). On top of that, it helps being more fancy and showing errors in red :nail_care:

**Comment by llucax** [comment]
@joshtriplett can you report this to openssh? In the meantime I don't really know what to do, because even if it weren't really necessary to detect git errors more reliably, I don't think there should be anything wrong with capturing stderr really, and this is somehow an obscure use case.

**Comment by joshtriplett** [comment]
@llucax I'm not sure if this is an SSH bug; it does normally print errors to the terminal even from the master-mode process.

I think there's a way to work around this in git-hub.  You want to wait for the process you spawned to exit, not any subprocesses it spawns; once the process you spawned exits, do one last non-blocking read of stderr, then proceed.

**Comment by jwilk** [comment]
This will be fixed in OpenSSH 7.3:
https://bugzilla.mindrot.org/show_bug.cgi?id=1988
https://github.com/openssh/openssh-portable/commit/d2d6bf864e52

**Comment by joshtriplett** [comment]
@jwilk I stand corrected. Great to see, thank you.
